internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+//
+// On success, a callback is returned which should be called to release the lock.
+func acquireSharedLock(path string, mode LockMode) (release func(), err error) {
+	// Initially open the device or file for reading
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot open device: %w", err)
+	}
+	defer f.Close()
+
+	how := unix.LOCK_SH
+	if mode == LockModeNonBlocking {
+		how |= unix.LOCK_NB
+	}
+
+	// Obtain information about the opened device or file
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot obtain file info: %w", err)
+	}
+
+	var lockPath string
+	var openFlags int
+
+	switch {
+	case isBlockDevice(fi.Mode()):
+		// Block devices use an advisory lock on a file in /run/cryptsetup. The lock file
+		// filename is of the format "L_<major>:<minor>".
+
+		// Don't assume that the lock directory exists.
+		if err := os.Mkdir(cryptsetupLockDir(), 0700); err != nil && !os.IsExist(err) {
+			return nil, xerrors.Errorf("cannot create lock directory: %w", err)
+		}
+
+		// Obtain information about the opened block device using the fstat syscall,
+		// where we get more information.
+		var st unix.Stat_t
+		if err := dataDeviceFstat(int(f.Fd()), &st); err != nil {
+			return nil, xerrors.Errorf("cannot obtain device info: %w", err)
+		}
+		lockPath = filepath.Join(cryptsetupLockDir(), fmt.Sprintf("L_%d:%d", unix.Major(st.Rdev), unix.Minor(st.Rdev)))
+		openFlags = os.O_RDWR | os.O_CREATE
+	case fi.Mode().IsRegular():
+		// Regular files use an advisory lock directly on the file.
+		lockPath = path
+		openFlags = os.O_RDWR
+	default:
+		return nil, errors.New("unsupported file type")
+	}
+
+	for {
+		// Attempt to open the lock file for writing.
+		lockFile, err := os.OpenFile(lockPath, openFlags, 0600)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot open lock file for writing: %w", err)
+		}
+
+		succeeded := false
+		defer func() {
+			if succeeded {
+				return
+			}
+			lockFile.Close()
+		}()
+
+		// Obtain and save information about the opened lock file.
+		var origSt unix.Stat_t
+		if err := unix.Fstat(int(lockFile.Fd()), &origSt); err != nil {
+			return nil, xerrors.Errorf("cannot obtain lock file info: %w", err)
+		}
+
+		// Attempt to acquire the requested lock.
+		if err := unix.Flock(int(lockFile.Fd()), how); err != nil {
+			return nil, xerrors.Errorf("cannot obtain lock: %w", err)
+		}
+
+		if isBlockDevice(fi.Mode()) {
+			// If we are attempting to acquire a lock on a block device, make sure that we
+			// aren't racing with a previous lock holder or a new lock holder.
+			//
+			// Obtain the information about the file currently linked from the lock file path.
+			var st unix.Stat_t
+			if err := unix.Stat(lockPath, &st); err != nil {
+				// The lock file we opened was unlinked by another lock owner between us
+				// opening the file and acquiring the lock. We need to try again.
+				lockFile.Close()
+				continue
```

Comment:
it's probably slightly hard to setup but this continue and the next are not tested